### ISSUE TITLE
Associated values in enum makes the use of old Matcher API instead of Predicate.

### ISSUE THREAD

**Issue by wrutkowski** [Quick/Nimble#834]
- [x] I have read [CONTRIBUTING](https://github.com/Quick/Nimble/blob/master/CONTRIBUTING.md) and have done my best to follow them.

### What did you do?

```swift
enum MyEnum: Equatable {
    case firstCase
    case secondCase
    case thirdCase(value: Int)
}

expect(MyEnum.firstCase).to(equal(.secondCase)) // this is fine
expect(MyEnum.firstCase).to(equal(.thirdCase(value: 10))) // this is also fine
var optionalEnum: MyEnum?
optionalEnum = .firstCase
expect(optionalEnum).to(equal(.secondCase)) // this is fine
expect(optionalEnum).to(equal(.thirdCase(value: 10))) // this produces warning that we are using old API ('to(_:description:)' is deprecated: Use Predicate instead)
```

When expected value is of an optional type and matcher value uses enum with associated value the old API is used.

### What did you expect to happen?

`expect(optionalEnum).to(equal(.thirdCase(value: 10)))` should create `Predicate` instead of a Matcher.

### What actually happened instead?

`Matcher` was used resulting in deprecated warning.

### Environment

List the software versions you're using:

- Nimble: 9.0.0-rc.3
- Xcode Version: 12.0
- Swift Version: 5.3

Please also mention which package manager you used and its version. Delete the
other package managers in this list:

- Cocoapods: 1.9.3

### Project that demonstrates the issue

Example code above.

**Comment by ikesyo** [comment]
I'm really not sure why Swift 5.3 compiler pick [this](https://github.com/Quick/Nimble/blob/1dc243993053d34f857aec54f2f5db520f34ff82/Sources/Nimble/Expectation.swift#L72) up only for such specific case 🤔

**Comment by wrutkowski** [comment]
I think you are right. Even providing the type explicitly fixes the warning:
```
expect(optionalEnum).to(equal(MyEnum.thirdCase(value: 10)))
```

**Comment by mseijas** [comment]
Seeing the same problem too!

**Comment by Robert-Hillary** [comment]
I'm also seeing this issue.

**Comment by stefanrenne** [comment]
I Just created a patch for this issue, let's hope the maintainers will have a look at it ;)

**Comment by stefanrenne** [comment]
@ikesyo any updates on this issue? When can we expect a major version bump?

**Comment by ratkins** [comment]
Getting something similar here in Xcode 12.3:
```
let actualCommentIds: [Comment] = // ...
expect(actualCommentIds).toEventually(equal([comment.id]))
```
Something to do with the automatic conformance of `Array<Equatable>` to `Equatable`?

**Comment by ikesyo** [comment]
I tested this on Xcode 12.5 and doesn't happen anymore. Looks like the overload resolution is fixed/improved in Swift 5.4. Closing.
